File: packages/gecko_db/lib/src/wire/format_header.dart
/// Versioned on-disk/wire compatibility header (Phase 0 contract).
///
/// This header is deliberately independent of the native worker so the Dart
/// side can validate artifacts before handing bytes to FFI. The native worker
/// must emit/consume the same fixed layout before a file is opened.
library;

import 'dart:convert';
import 'dart:typed_data';

import '../errors/errors.dart';

/// Current on-disk format version.
const int geckoFormatVersion = 1;

/// Current Dart/native wire protocol version.
const int geckoWireVersion = 1;

/// A deterministic compatibility header.
class FormatHeader {
  const FormatHeader({
    this.formatVersion = geckoFormatVersion,
    this.wireVersion = geckoWireVersion,
    required this.packageVersion,
  });

  static const List<int> magic = <int>[0x47, 0x45, 0x43, 0x4B, 0x4F, 0x01];

  final int formatVersion;
  final int wireVersion;
  final String packageVersion;

  Uint8List encode() {
    final version = utf8.encode(packageVersion);
    if (version.length > 255 ||
        formatVersion < 0 ||
        formatVersion > 255 ||
        wireVersion < 0 ||
        wireVersion > 255) {
      throw GeckoError(
        GeckoErrorType.invalidOperation,
        'Format header values are outside the supported wire range',
      );
    }
    return Uint8List.fromList(<int>[
      ...magic,
      formatVersion,
      wireVersion,
      version.length,
      ...version,
    ]);
  }

  static FormatHeader decode(List<int> bytes) {
    if (bytes.length < magic.length + 3 || !_startsWith(bytes, magic)) {
      throw GeckoError(
        GeckoErrorType.checksumMismatch,
        'Invalid gecko_db format header magic or truncated header',
      );
    }
    final packageLength = bytes[magic.length + 2];
    final expectedLength = magic.length + 3 + packageLength;
    if (bytes.length != expectedLength) {
      throw GeckoError(
        GeckoErrorType.checksumMismatch,
        'Invalid gecko_db format header length',
      );
    }
    final formatVersion = bytes[magic.length];
    final wireVersion = bytes[magic.length + 1];
    final packageVersion = utf8.decode(bytes.sublist(magic.length + 3));
    return FormatHeader(
      formatVersion: formatVersion,
      wireVersion: wireVersion,
      packageVersion: packageVersion,
    );
  }

  /// Throws `upgradeRequired` when the header is not readable by this client.
  void validateCompatibility({
    int expectedFormatVersion = geckoFormatVersion,
    int expectedWireVersion = geckoWireVersion,
  }) {
    if (formatVersion != expectedFormatVersion ||
        wireVersion != expectedWireVersion) {
      throw GeckoError(
        GeckoErrorType.upgradeRequired,
        'Database format $formatVersion/wire $wireVersion requires an upgrade '
        '(expected format $expectedFormatVersion/wire $expectedWireVersion)',
        details: <String, Object?>{
          'formatVersion': formatVersion,
          'wireVersion': wireVersion,
          'expectedFormatVersion': expectedFormatVersion,
          'expectedWireVersion': expectedWireVersion,
        },
      );
    }
  }

  static bool _startsWith(List<int> bytes, List<int> prefix) {
    if (bytes.length < prefix.length) return false;
    for (var i = 0; i < prefix.length; i++) {
      if (bytes[i] != prefix[i]) return false;
    }
    return true;
  }
}

File: rust/src/format_header.rs
//! Fixed on-disk/wire compatibility header mirroring Dart `FormatHeader`.

pub const FORMAT_VERSION: u8 = 1;
pub const WIRE_VERSION: u8 = 1;
pub const MAGIC: [u8; 6] = [0x47, 0x45, 0x43, 0x4B, 0x4F, 0x01];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FormatHeader {
    pub format_version: u8,
    pub wire_version: u8,
    pub package_version: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HeaderError(pub String);

impl std::fmt::Display for HeaderError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "HeaderError: {}", self.0)
    }
}
impl std::error::Error for HeaderError {}

type Result<T> = std::result::Result<T, HeaderError>;

impl FormatHeader {
    pub fn encode(&self) -> Result<Vec<u8>> {
        let version = self.package_version.as_bytes();
        if version.len() > u8::MAX as usize {
            return Err(HeaderError("package version is too long".into()));
        }
        let mut out = MAGIC.to_vec();
        out.push(self.format_version);
        out.push(self.wire_version);
        out.push(version.len() as u8);
        out.extend_from_slice(version);
        Ok(out)
    }

    pub fn decode(bytes: &[u8]) -> Result<Self> {
        if bytes.len() < MAGIC.len() + 3 || bytes[..MAGIC.len()] != MAGIC {
            return Err(HeaderError("invalid magic or truncated header".into()));
        }
        let package_len = bytes[MAGIC.len() + 2] as usize;
        let expected = MAGIC.len() + 3 + package_len;
        if bytes.len() != expected {
            return Err(HeaderError("invalid header length".into()));
        }
        let package_version = std::str::from_utf8(&bytes[MAGIC.len() + 3..])
            .map_err(|_| HeaderError("invalid package version UTF-8".into()))?
            .to_owned();
        Ok(Self {
            format_version: bytes[MAGIC.len()],
            wire_version: bytes[MAGIC.len() + 1],
            package_version,
        })
    }

    pub fn validate_compatibility(&self, format: u8, wire: u8) -> Result<()> {
        if self.format_version != format || self.wire_version != wire {
            return Err(HeaderError(format!(
                "requires format {}/wire {}, found {}/{}",
                format, wire, self.format_version, self.wire_version
            )));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let header = FormatHeader {
            format_version: FORMAT_VERSION,
            wire_version: WIRE_VERSION,
            package_version: "0.0.1".into(),
        };
        assert_eq!(
            FormatHeader::decode(&header.encode().unwrap()).unwrap(),
            header
        );
    }

    #[test]
    fn rejects_bad_header() {
        assert!(FormatHeader::decode(&[0, 1, 2]).is_err());
    }

    #[test]
    fn rejects_incompatible_versions() {
        let header = FormatHeader {
            format_version: FORMAT_VERSION,
            wire_version: WIRE_VERSION,
            package_version: "x".into(),
        };
        assert!(header.validate_compatibility(1, 99).is_err());
    }
}
